File: src/lib/logger.ts
import path from 'path';
import fs from 'fs';

const LOG_LEVELS = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  error: 4,
};

type LogLevel = keyof typeof LOG_LEVELS;

const currentLevel = (process.env.LOG_LEVEL?.toLowerCase() as LogLevel) || 'info';
const currentLevelValue = LOG_LEVELS[currentLevel] ?? 2;

const isProd = process.env.NODE_ENV === 'production';

function ensureDirectoryExists(filePath: string) {
  try {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  } catch (e) {
    console.error('Failed to create log directory:', e);
  }
}

function writeLog(level: LogLevel, message: string, meta?: Record<string, unknown>) {
  const levelVal = LOG_LEVELS[level];
  if (levelVal < currentLevelValue) return;

  const timestamp = new Date().toISOString();
  let logOutput = '';

  if (isProd) {
    // Structured JSON log output for Production
    logOutput = JSON.stringify({
      timestamp,
      level: level.toUpperCase(),
      message,
      ...meta,
    });
  } else {
    // Colorized/pretty output for Dev/QA
    const colors = {
      trace: '\x1b[90m', // Gray
      debug: '\x1b[36m', // Cyan
      info: '\x1b[32m',  // Green
      warn: '\x1b[33m',  // Yellow
      error: '\x1b[31m', // Red
      reset: '\x1b[0m',
    };
    const color = colors[level] || colors.reset;
    const metaStr = meta ? ` ${JSON.stringify(meta)}` : '';
    logOutput = `[${timestamp}] ${color}${level.toUpperCase()}${colors.reset}: ${message}${metaStr}`;
  }

  // Console output
  if (level === 'error') {
    console.error(logOutput);
  } else if (level === 'warn') {
    console.warn(logOutput);
  } else {
    console.log(logOutput);
  }

  // File logging (if LOG_PATH is configured in the environment)
  const logPath = process.env.LOG_PATH;
  if (logPath) {
    try {
      ensureDirectoryExists(logPath);
      fs.appendFileSync(logPath, (isProd ? logOutput : logOutput.replace(/\x1b\[[0-9;]*m/g, '')) + '\n', 'utf-8');
    } catch (e) {
      console.error('Failed to write log to file:', e);
    }
  }
}

export const logger = {
  trace: (msg: string, meta?: Record<string, unknown>) => writeLog('trace', msg, meta),
  debug: (msg: string, meta?: Record<string, unknown>) => writeLog('debug', msg, meta),
  info: (msg: string, meta?: Record<string, unknown>) => writeLog('info', msg, meta),
  warn: (msg: string, meta?: Record<string, unknown>) => writeLog('warn', msg, meta),
  error: (msg: string, meta?: Record<string, unknown>) => writeLog('error', msg, meta),
};

File: backend/app/core/logger.py
import logging
import json
import os
import sys
from datetime import datetime

# Map log level names to logging constants
LOG_LEVELS = {
    "trace": logging.DEBUG,
    "debug": logging.DEBUG,
    "info": logging.INFO,
    "warn": logging.WARNING,
    "error": logging.ERROR,
    "critical": logging.CRITICAL
}

log_level_str = os.getenv("LOG_LEVEL", "info").lower()
log_level = LOG_LEVELS.get(log_level_str, logging.INFO)

is_prod = os.getenv("NODE_ENV") == "production"

class CustomJSONFormatter(logging.Formatter):
    def format(self, record):
        log_data = {
            "timestamp": datetime.utcfromtimestamp(record.created).isoformat() + "Z",
            "level": record.levelname,
            "message": record.getMessage()
        }
        if record.exc_info:
            log_data["exception"] = self.formatException(record.exc_info)
        if hasattr(record, "extra_data"):
            log_data.update(record.extra_data)
        return json.dumps(log_data)

class ColorFormatter(logging.Formatter):
    COLORS = {
        "DEBUG": "\033[36m",    # Cyan
        "INFO": "\033[32m",     # Green
        "WARNING": "\033[33m",  # Yellow
        "ERROR": "\033[31m",    # Red
        "CRITICAL": "\033[41m\033[37m" # Red background, white text
    }
    RESET = "\033[0m"

    def format(self, record):
        timestamp = datetime.utcfromtimestamp(record.created).isoformat() + "Z"
        level = record.levelname
        color = self.COLORS.get(level, self.RESET)
        message = record.getMessage()
        exc = ""
        if record.exc_info:
            exc = f"\n{self.formatException(record.exc_info)}"
        return f"[{timestamp}] {color}{level}{self.RESET}: {message}{exc}"

def get_logger(name="phantom"):
    logger = logging.getLogger(name)
    logger.setLevel(log_level)
    
    # Avoid duplicate handlers if already configured
    if logger.handlers:
        return logger
        
    # Console Handler
    ch = logging.StreamHandler(sys.stdout)
    ch.setLevel(log_level)
    if is_prod:
        ch.setFormatter(CustomJSONFormatter())
    else:
        ch.setFormatter(ColorFormatter())
    logger.addHandler(ch)

    # File Handler (if LOG_PATH is defined)
    log_path = os.getenv("LOG_PATH")
    if log_path:
        try:
            os.makedirs(os.path.dirname(log_path), exist_ok=True)
            fh = logging.FileHandler(log_path, encoding="utf-8")
            fh.setLevel(log_level)
            if is_prod:
                fh.setFormatter(CustomJSONFormatter())
            else:
                class ColorFreeFormatter(logging.Formatter):
                    def format(self, record):
                        timestamp = datetime.utcfromtimestamp(record.created).isoformat() + "Z"
                        level = record.levelname
                        message = record.getMessage()
                        exc = ""
                        if record.exc_info:
                            exc = f"\n{self.formatException(record.exc_info)}"
                        return f"[{timestamp}] {level}: {message}{exc}"
                fh.setFormatter(ColorFreeFormatter())
            logger.addHandler(fh)
        except Exception as e:
            print(f"Failed to setup file logging: {e}", file=sys.stderr)
            
    return logger

logger = get_logger()
